kalphitequeen/src/main/java/net/runelite/client/plugins/kalphitequeen/KQPlugin.java

```diff
@@ -0,0 +1,241 @@
+
+	@Getter
+	private KQNpc kalphiteQueen;
+
+	private boolean atKalphiteQueen;
+
+	@Getter
+	@Setter
+	private boolean flashVeng;
+
+	private int lastVengActive;
+	private int lastVengCooldown;
+
+	@Provides
+	KQConfig getConfig(final ConfigManager configManager)
+	{
+		return configManager.getConfig(KQConfig.class);
+	}
+
+	@Override
+	protected void startUp()
+	{
+		if (client.getGameState() != GameState.LOGGED_IN || !atKalphiteQueen())
+		{
+			return;
+		}
+
+		init();
+	}
+
+	@Override
+	protected void shutDown()
+	{
+		eventBus.unregister(this);
+
+		overlayManager.remove(sceneOverlay);
+
+		atKalphiteQueen = false;
+
+		flashVeng = false;
+		lastVengActive = 0;
+		lastVengCooldown = 0;
+
+		kalphiteQueen = null;
+	}
+
+	@Subscribe
+	private void onGameStateChanged(final GameStateChanged event)
+	{
+		final GameState gameState = event.getGameState();
+
+		switch (gameState)
+		{
+			case LOGGED_IN:
+				if (atKalphiteQueen())
+				{
+					if (!atKalphiteQueen)
+					{
+						init();
+					}
+				}
+				else
+				{
+					if (atKalphiteQueen)
+					{
+						shutDown();
+					}
+				}
+				break;
+			case LOGIN_SCREEN:
+			case HOPPING:
+				if (atKalphiteQueen)
+				{
+					shutDown();
+				}
+				break;
+		}
+	}
+
+	private void onGameTick(final GameTick event)
+	{
+		if (kalphiteQueen != null)
+		{
+			kalphiteQueen.updateTicksUntilNextAnimation();
+		}
+	}
+
+	private void onVarbitChanged(final VarbitChanged event)
+	{
+		final int vengCoolDown = client.getVar(Varbits.VENGEANCE_COOLDOWN);
+		final int vengActive = client.getVar(Varbits.VENGEANCE_ACTIVE);
+
+		if (lastVengCooldown != vengCoolDown)
+		{
+			if (vengCoolDown == 0 && vengActive == 0)
+			{
+				flashVeng = true;
+			}
+
+			lastVengCooldown = vengCoolDown;
+		}
+
+		if (lastVengActive != vengActive)
+		{
+			if (vengActive == 0 && vengCoolDown == 0)
+			{
+				flashVeng = true;
+			}
+
+			lastVengActive = vengActive;
+		}
+	}
+
+	private void onNpcSpawned(final NpcSpawned event)
+	{
+		final NPC npc = event.getNpc();
+
+		addNpc(npc);
+	}
+
+	private void onNpcDespawned(final NpcDespawned event)
+	{
+		final NPC npc = event.getNpc();
+
+		removeNpc(npc);
+	}
+
+	private void addNpc(final NPC npc)
+	{
+		if (KQNpc.NPC_IDS.contains(npc.getId()))
+		{
+			kalphiteQueen = new KQNpc(npc);
+		}
+	}
+
+	private void removeNpc(final NPC npc)
+	{
+		if (KQNpc.NPC_IDS.contains(npc.getId()))
+		{
+			kalphiteQueen = null;
+		}
+	}
+
+	private void init()
+	{
+		atKalphiteQueen = true;
+
+		overlayManager.add(sceneOverlay);
+
+		for (final NPC npc : client.getNpcs())
+		{
+			addNpc(npc);
+		}
+
+		eventBus.subscribe(GameTick.class, this, this::onGameTick);
+		eventBus.subscribe(NpcSpawned.class, this, this::onNpcSpawned);
+		eventBus.subscribe(NpcDespawned.class, this, this::onNpcDespawned);
+		eventBus.subscribe(VarbitChanged.class, this, this::onVarbitChanged);
+	}
+
+	private boolean atKalphiteQueen()
+	{
+		return client.getMapRegions()[0] == KQ_REGION;
```

Comment:
this doesn't account for multiple scene planes. it should only account for the scene you are on.